substrate/pkg/controller/substrate/cluster/instanceprofile.go

```diff
@@ -41,115 +59,154 @@ var AssumeRolePolicyDocument = aws.String(`{
+	for _, policy := range managedPolicies {
+		if _, err := i.IAM.DetachRolePolicyWithContext(ctx, &iam.DetachRolePolicyInput{RoleName: resourceName, PolicyArn: aws.String(policy)}); err != nil {
 			if err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
 				return reconcile.Result{}, fmt.Errorf("detatching policy from role, %w", err)
 			}
 		} else {
-			logging.FromContext(ctx).Infof("Deleted policy %s from role %s", aws.StringValue(discovery.Name(substrate)), aws.StringValue(discovery.Name(substrate)))
+			logging.FromContext(ctx).Infof("Deleted policy %s from role %s", aws.StringValue(resourceName), aws.StringValue(resourceName))
 		}
 	}
 	// Binding
-	if _, err := i.IAM.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{RoleName: discovery.Name(substrate), InstanceProfileName: discovery.Name(substrate)}); err != nil {
+	if _, err := i.IAM.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{RoleName: resourceName, InstanceProfileName: resourceName}); err != nil {
 		if err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
 			return reconcile.Result{}, fmt.Errorf("removing instance profile from role %w,", err)
 		}
 	} else {
-		logging.FromContext(ctx).Infof("Deleted role %s from instance profile %s", aws.StringValue(discovery.Name(substrate)), aws.StringValue(discovery.Name(substrate)))
+		logging.FromContext(ctx).Infof("Deleted role %s from instance profile %s", aws.StringValue(resourceName), aws.StringValue(resourceName))
 	}
 	// Profile
-	if _, err := i.IAM.DeleteInstanceProfileWithContext(ctx, &iam.DeleteInstanceProfileInput{InstanceProfileName: discovery.Name(substrate)}); err != nil {
+	if _, err := i.IAM.DeleteInstanceProfileWithContext(ctx, &iam.DeleteInstanceProfileInput{InstanceProfileName: resourceName}); err != nil {
 		if err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
 			return reconcile.Result{}, fmt.Errorf("deleting instance profile %w,", err)
 		}
 	} else {
-		logging.FromContext(ctx).Infof("Deleted instance profile %s", aws.StringValue(discovery.Name(substrate)))
+		logging.FromContext(ctx).Infof("Deleted instance profile %s", aws.StringValue(resourceName))
 	}
 	// Role
-	if _, err := i.IAM.DeleteRoleWithContext(ctx, &iam.DeleteRoleInput{RoleName: discovery.Name(substrate)}); err != nil {
+	if _, err := i.IAM.DeleteRoleWithContext(ctx, &iam.DeleteRoleInput{RoleName: resourceName}); err != nil {
 		if err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
 			return reconcile.Result{}, fmt.Errorf("deleting role, %w", err)
 		}
 	} else {
-		logging.FromContext(ctx).Infof("Deleted role %s", aws.StringValue(discovery.Name(substrate)))
+		logging.FromContext(ctx).Infof("Deleted role %s", aws.StringValue(resourceName))
 	}
 	return reconcile.Result{}, nil
 }
+
+func desiredRolesFor(substrate *v1alpha1.Substrate) []roleInfo {
```

Comment:
scope this function to the instanceprofile controller

func (i *InstanceProfile) roles(substrate *v1alpha1.Substrate)

Reply:
Is there a benefit in doing this? I usually try to avoid having a receiver if not required.